bus/contractlocking.go

```diff
@@ -0,0 +1,205 @@
+// contractLocks.Acquire is closed before the lock can be acquired.
+var ErrAcquireContractTimeout = errors.New("acquiring the lock timed out")
+
+type contractLocks struct {
+	mu    sync.Mutex
+	locks map[types.FileContractID]*contractLock
+}
+
+type contractLock struct {
+	mu          sync.Mutex    // locks contractLock fields
+	lockChan    chan struct{} // locks the contract
+	lockedUntil time.Time
+	heldBy      uint64
+	queue       []*lockCandidate
+}
+
+type lockCandidate struct {
+	lockID       uint64
+	lockDuration time.Duration
+	c            chan struct{}
+}
+
+func newContractLocks() *contractLocks {
+	return &contractLocks{
+		locks: make(map[types.FileContractID]*contractLock),
+	}
+}
+
+func (l *contractLocks) lockForContractID(id types.FileContractID, create bool) *contractLock {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	lock, exists := l.locks[id]
+	if !exists && create {
+		c := make(chan struct{})
+		close(c)
+		lock = &contractLock{
+			lockChan:    c,
+			lockedUntil: time.Now().Add(time.Hour),
+		}
+		l.locks[id] = lock
+	}
+	return lock
+}
+
+// Acquire acquires a contract lock for the given id and provided duration. If
+// acquiring the lock doesn't finish before the context is closed,
+// ErrAcquireContractTimeout is returned. Upon success an identifier is returned
+// which can be used to release the lock before its lock duration has passed.
+// TODO: Extend this with some sort of priority. e.g. migrations would acquire a
+// lock with a low priority but contract maintenance would have a very high one
+// to avoid being starved by low prio tasks.
+func (l *contractLocks) Acquire(ctx context.Context, id types.FileContractID, d time.Duration) (uint64, error) {
+	lock := l.lockForContractID(id, true)
+
+	// Prepare a random lockID for ourselves.
+	ourLockID := frand.Uint64n(math.MaxUint64)
+
+	// Prepare a channel to indicate to other candidates whether we are
+	// still waiting to acquire the lock.
+	done := make(chan struct{})
+	defer close(done)
+
+	// Add ourselves to the queue of candidates for locking, grab the
+	// lockChan to be notified when the lock is released prematurely and
+	// prepare a timer to be notified when the current lock expires.
+	lock.mu.Lock()
+	lockChan := lock.lockChan
+	lock.queue = append(lock.queue, &lockCandidate{
+		lockID:       ourLockID,
+		lockDuration: d,
+		c:            done,
+	})
+	t := time.NewTimer(time.Until(lock.lockedUntil)) // prepare a timer for expiring locks
+	lock.mu.Unlock()
+
+	// Drain timer at the end if necessary.
+	drainTimer := func() {
+		if !t.Stop() {
+			select {
+			case <-t.C:
+			default:
+			}
+		}
+	}
+	defer drainTimer()
+
+	// tryAcquire is a helper function that tries to acquire a lock
+	// whenenver waiting threads are either woken by the lock timing out or
+	// the lock being released.
+	tryAcquire := func() (uint64, bool, error) {
+		lock.mu.Lock()
+
+		// Fetch the next candidate for locking. Drop any that
```

Comment:
I'm wondering whether the next candidate shoudn't be unlocked by a goroutine that's spun from the release. When I thought about acquiring locks and queueing and stuff I figured an acquire needs to have a happy path: 
- acquire lock
- if first on there: great - return
- if needs to wait: enqueue yourself - block on your own channel and the lock timer

Then the release should
- release lock
- go through the queue and unblocks the next in line

Perhaps we can update `lockForContractID` to return whether it created the lock, then `Acquire` has a happy flow at least where it doesn' thave to care about queueing and candidates. I don't mind the current implementation, the more I think about it the more I see why it's not trivial, decided the keep the comment as is because maybe it does spark something you want to implement/change.